File: hphp/hack/src/hhbc/iterator.ml
(*
 * Copyright (c) 2017, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

type t = Id of int

let to_string (Id i) = string_of_int i

let next_iterator = ref 0

let num_iterators = ref 0

let get_iterator () =
  let current = !next_iterator in
  next_iterator := current + 1;
  num_iterators := max !num_iterators !next_iterator;
  Id current

let free_iterator () = next_iterator := !next_iterator - 1

let reset_iterator () =
  next_iterator := 0;
  num_iterators := 0

File: hphp/hack/src/hhbc/iterator.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use std::fmt;

#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
pub struct Id(pub usize);

impl fmt::Display for Id {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

#[derive(Default, Debug, Clone)]
pub struct Iter {
    pub next: Id,
    count: usize,
}
impl Iter {
    pub fn count(&self) -> usize {
        self.count
    }

    pub fn get(&mut self) -> Id {
        let curr = self.next.0;
        self.next.0 = curr + 1;
        self.count = usize::max(self.count, self.next.0);
        Id(curr)
    }

    pub fn free(&mut self) {
        self.next.0 -= 1;
    }

    pub fn reset(&mut self) {
        *self = Self::default();
    }
}
